File: rs2-core/src/crypto.rs
//! Small shared crypto helpers (hex codecs + keyed/unkeyed SHA hashing) used
//! by the pipeline transform crypto functions (`$hmac`/`$hmacVerify`) and by
//! outbound credential injection (HMAC/AWS SigV4 signing). Kept in one place so
//! the two callers don't drift on algorithm support or hex formatting.

use hmac::{Hmac, Mac};
use sha2::{Digest, Sha256, Sha512};

/// Lowercase-hex encode bytes.
pub fn to_hex(bytes: &[u8]) -> String {
    const HEX: &[u8; 16] = b"0123456789abcdef";
    let mut out = String::with_capacity(bytes.len() * 2);
    for b in bytes {
        out.push(HEX[(b >> 4) as usize] as char);
        out.push(HEX[(b & 0xf) as usize] as char);
    }
    out
}

/// Decode lowercase/uppercase hex; `None` on odd length or non-hex digits.
pub fn from_hex(s: &str) -> Option<Vec<u8>> {
    let s = s.trim();
    if s.is_empty() || !s.len().is_multiple_of(2) {
        return None;
    }
    (0..s.len())
        .step_by(2)
        .map(|i| u8::from_str_radix(&s[i..i + 2], 16).ok())
        .collect()
}

/// HMAC of `message` under `key` for `sha256`/`sha512`; `None` on unknown algo.
pub fn hmac_bytes(algorithm: &str, key: &[u8], message: &[u8]) -> Option<Vec<u8>> {
    match algorithm {
        "sha256" => {
            let mut mac = Hmac::<Sha256>::new_from_slice(key).ok()?;
            mac.update(message);
            Some(mac.finalize().into_bytes().to_vec())
        }
        "sha512" => {
            let mut mac = Hmac::<Sha512>::new_from_slice(key).ok()?;
            mac.update(message);
            Some(mac.finalize().into_bytes().to_vec())
        }
        _ => None,
    }
}

/// Constant-time HMAC verification against a hex signature. Any malformed input
/// (unknown algorithm, non-hex signature) → `false`.
pub fn hmac_verify(algorithm: &str, key: &[u8], message: &[u8], signature_hex: &str) -> bool {
    let Some(provided) = from_hex(signature_hex) else {
        return false;
    };
    match algorithm {
        "sha256" => match Hmac::<Sha256>::new_from_slice(key) {
            Ok(mut mac) => {
                mac.update(message);
                mac.verify_slice(&provided).is_ok()
            }
            Err(_) => false,
        },
        "sha512" => match Hmac::<Sha512>::new_from_slice(key) {
            Ok(mut mac) => {
                mac.update(message);
                mac.verify_slice(&provided).is_ok()
            }
            Err(_) => false,
        },
        _ => false,
    }
}

/// HMAC-SHA256 raw bytes (the SigV4 signing primitive).
pub fn hmac_sha256(key: &[u8], message: &[u8]) -> Vec<u8> {
    let mut mac = Hmac::<Sha256>::new_from_slice(key).expect("HMAC accepts any key length");
    mac.update(message);
    mac.finalize().into_bytes().to_vec()
}

/// SHA-256 digest, lowercase hex (SigV4 payload/canonical-request hashing).
pub fn sha256_hex(bytes: &[u8]) -> String {
    let mut hasher = Sha256::new();
    hasher.update(bytes);
    to_hex(&hasher.finalize())
}

#[cfg(test)]
mod tests {
    use super::*;

    // RFC 4231 test case 2: key "Jefe", data "what do ya want for nothing?".
    #[test]
    fn hmac_sha256_known_vector() {
        let mac = hmac_bytes("sha256", b"Jefe", b"what do ya want for nothing?").unwrap();
        assert_eq!(
            to_hex(&mac),
            "5bdcc146bf60754e6a042426089575c75a003f089d2739839dec58b964ec3843"
        );
    }

    #[test]
    fn hex_roundtrips() {
        let bytes = vec![0u8, 1, 15, 16, 255];
        assert_eq!(from_hex(&to_hex(&bytes)).unwrap(), bytes);
        assert!(from_hex("xyz").is_none());
        assert!(from_hex("abc").is_none()); // odd length
    }

    // NIST FIPS 180-2: SHA-256 of "abc".
    #[test]
    fn sha256_known_vector() {
        assert_eq!(
            sha256_hex(b"abc"),
            "ba7816bf8f01cfea414140de5dae2223b00361a396177a9cb410ff61f20015ad"
        );
    }
}

File: rs2-worker/src/runtime/crypto.ts
// Small shared crypto helpers over WebCrypto. Port of `rs2-core/src/crypto.rs`.

const encoder = new TextEncoder();

/// Lowercase-hex encode bytes.
export function toHex(bytes: Uint8Array | ArrayBuffer): string {
  const b = bytes instanceof Uint8Array ? bytes : new Uint8Array(bytes);
  let out = "";
  for (const x of b) out += x.toString(16).padStart(2, "0");
  return out;
}

/// Decode lowercase/uppercase hex; `undefined` on odd length or non-hex digits.
export function fromHex(s: string): Uint8Array | undefined {
  const t = s.trim();
  if (t.length === 0 || t.length % 2 !== 0 || !/^[0-9a-fA-F]+$/.test(t)) return undefined;
  const out = new Uint8Array(t.length / 2);
  for (let i = 0; i < t.length; i += 2) out[i / 2] = parseInt(t.slice(i, i + 2), 16);
  return out;
}

function bytesOf(v: string | Uint8Array): Uint8Array {
  return typeof v === "string" ? encoder.encode(v) : v;
}

/// Copy into a fresh ArrayBuffer-backed view (WebCrypto rejects views over
/// SharedArrayBuffer / offset views in some runtimes).
function asBuffer(v: Uint8Array): ArrayBuffer {
  const out = new ArrayBuffer(v.byteLength);
  new Uint8Array(out).set(v);
  return out;
}

/// HMAC of `message` under `key` for `sha256`/`sha512`; `undefined` on an
/// unknown algorithm.
export async function hmacBytes(
  algorithm: string,
  key: string | Uint8Array,
  message: string | Uint8Array,
): Promise<Uint8Array | undefined> {
  const hash = algorithm === "sha256" ? "SHA-256" : algorithm === "sha512" ? "SHA-512" : undefined;
  if (!hash) return undefined;
  const k = await crypto.subtle.importKey("raw", asBuffer(bytesOf(key)), { name: "HMAC", hash }, false, ["sign"]);
  const sig = await crypto.subtle.sign("HMAC", k, asBuffer(bytesOf(message)));
  return new Uint8Array(sig);
}

/// Constant-time HMAC verification against a hex signature.
export async function hmacVerify(
  algorithm: string,
  key: string | Uint8Array,
  message: string | Uint8Array,
  signatureHex: string,
): Promise<boolean> {
  const provided = fromHex(signatureHex);
  if (!provided) return false;
  const expected = await hmacBytes(algorithm, key, message);
  if (!expected) return false;
  return constantTimeEqual(expected, provided);
}

/// HMAC-SHA256 raw bytes (the SigV4 signing primitive).
export async function hmacSha256(key: string | Uint8Array, message: string | Uint8Array): Promise<Uint8Array> {
  return (await hmacBytes("sha256", key, message))!;
}

/// SHA-256 digest bytes.
export async function sha256(bytes: string | Uint8Array): Promise<Uint8Array> {
  return new Uint8Array(await crypto.subtle.digest("SHA-256", asBuffer(bytesOf(bytes))));
}

/// SHA-256 digest, lowercase hex.
export async function sha256Hex(bytes: string | Uint8Array): Promise<string> {
  return toHex(await sha256(bytes));
}

/// Constant-time byte equality. Unequal lengths fail fast — the length isn't
/// the secret.
export function constantTimeEqual(a: Uint8Array, b: Uint8Array): boolean {
  if (a.byteLength !== b.byteLength) return false;
  const subtle = crypto.subtle as unknown as { timingSafeEqual?: (a: ArrayBuffer, b: ArrayBuffer) => boolean };
  if (typeof subtle.timingSafeEqual === "function") {
    return subtle.timingSafeEqual(asBuffer(a), asBuffer(b));
  }
  let diff = 0;
  for (let i = 0; i < a.byteLength; i++) diff |= a[i]! ^ b[i]!;
  return diff === 0;
}

export function base64Encode(bytes: Uint8Array): string {
  let s = "";
  for (const b of bytes) s += String.fromCharCode(b);
  return btoa(s);
}

export function base64Decode(s: string): Uint8Array {
  const bin = atob(s);
  const out = new Uint8Array(bin.length);
  for (let i = 0; i < bin.length; i++) out[i] = bin.charCodeAt(i);
  return out;
}

export function base64UrlEncode(bytes: Uint8Array): string {
  return base64Encode(bytes).replace(/\+/g, "-").replace(/\//g, "_").replace(/=+$/, "");
}

/// Strict base64url (no padding) decode; `undefined` on malformed input.
export function base64UrlDecode(s: string): Uint8Array | undefined {
  if (!/^[A-Za-z0-9_-]*$/.test(s)) return undefined;
  const pad = s.length % 4;
  if (pad === 1) return undefined;
  const b64 = s.replace(/-/g, "+").replace(/_/g, "/") + "=".repeat(pad === 0 ? 0 : 4 - pad);
  try {
    return base64Decode(b64);
  } catch {
    return undefined;
  }
}
